Extra change for Moonfire NVR Web UI

michioxd opened the pull request:
![image](https://github.com/scottlamb/moonfire-nvr/assets/80969068/ca437880-7b94-4d93-b84c-8515466563a5)


This commit has changed/added those feature below:

- Fixed lagging on drawer open/close by separating the drawer and header into separate components.
- Fixed filter layout (used `flex`).
- Added dark theme support (including system theme, dark, light).
- Added outlined filter button if it enabled.
 
  ![image](https://github.com/scottlamb/moonfire-nvr/assets/80969068/b1b545d7-7323-47dd-9315-b7331dbe6f1c)

  ![image](https://github.com/scottlamb/moonfire-nvr/assets/80969068/bee96f26-ef37-44a2-863d-975ad004426d)

- Redesign login dialog.
  
  ![image](https://github.com/scottlamb/moonfire-nvr/assets/80969068/5bd4c1e7-8cbb-4427-b1c9-c3d59bbae562)

- Added dual camera layout.
- Each layout now has a name instead of a class name to display in the layout selector.
- Added toggle full screen button.
- Added save current layout and opened cameras to `localStorage`.

  ![image](https://github.com/scottlamb/moonfire-nvr/assets/80969068/b444dc48-8e8f-4c35-86bf-e16cd31b440b)

- Added support for legacy browser (see [`vite.config.ts`](https://github.com/scottlamb/moonfire-nvr/pull/315/files#diff-3ac010a0a7c98e20e2a93940f788480b80115f74ebbe72c7254b3e3bcdbc3c8e))

michioxd commented:
oh damn I used `pnpm` instead `npm` heh...

scottlamb commented on ui/src/List/DisplaySelector.tsx:
```diff
@@ -33,15 +33,14 @@ export const DEFAULT_DURATION = DURATIONS[0][1];
         flexDirection: "column",
       }}
     >
+      <CardContent>
```
Ahh, I guess missing this was why the `padding: theme.spacing(1)` was necessary before above?

Could you fix the indentation within this element?

michioxd replied:
We don't need padding for this because the `<CardContent />` has done it for us. See full `Card` usage in the [MUI Documentation](https://mui.com/material-ui/react-card/#basics).

scottlamb replied:
So from the docs, looks like you're totally right that `Card` should have a `CardContent` within it.

But I noticed this morning that it spaces out the UI enough to introduce a vertical scroll bar even on my large desktop monitor; `Card` before vs `CardContent` after:

<img width="208" alt="image" src="https://github.com/scottlamb/moonfire-nvr/assets/8949524/bf3a9e2c-c867-4416-82a2-a799ef354e2b"> <img width="201" alt="image" src="https://github.com/scottlamb/moonfire-nvr/assets/8949524/530cc149-f258-48a3-9c45-9baa86c023bc">

* 12 extra pixels of interior width, I think because one or more of the cards in the column no longer fits quite as well in the column
* 8+8 => 16+16 pixels of horizontal padding
* 8+8 => 16+24 pixels of vertical padding
* margin went away, but I think that just turned into the same space showing up in a different way between elements

Looks like the padding comes from here:

https://github.com/mui/material-ui/blob/9ef12dc81c54566eeceae16bcf026a75fbce504d/packages/mui-material/src/CardContent/CardContent.js#L21-L30

All the extra spacing might be my biggest complaint in general about the mui library, e.g. <https://github.com/mui/material-ui/issues/27700>.

Looks like I could re-introduce the `sx={{ padding: theme.spacing(1) }}` on the `CardContent` and get it closer to before, but I'm not sure how to best undo the `last-child` bit.

I don't really remember why I picked `Card` to begin with, though. I just wanted something to group like controls but my usage doesn't really feel like their examples. Should it be just `Paper` instead? `<Paper sx={{ padding: theme.spacing(1) }}>` seems to render as before (with the margin going somewhere else I guess):

<img width="195" alt="image" src="https://github.com/scottlamb/moonfire-nvr/assets/8949524/bb261f62-5a96-4083-95a3-ffbe97f63a8b">

michioxd replied:
Hmm, I think that padding at `last-child` should separate between `CardContent` and `CardActions` (see [Material Design 2 Guidelines](https://m2.material.io/components/cards#specs)) but we don't have a plan to use `CardActions`. Maybe we can override it by adding `sx={{padding: 1}}`

scottlamb replied:
With `sx={{padding: 1}}` the `last-child` thing is still there. iiuc it's more selective so it wins out.

<img width="690" alt="image" src="https://github.com/scottlamb/moonfire-nvr/assets/8949524/be4e062f-72b5-40cf-b952-2652b84ee2ec">

> 12 extra pixels of interior width, I think because one or more of the cards in the column no longer fits quite as well in the column

Huh. By commenting stuff out, I narrowed it down to the "Trim start and end" in the display selector. And it's fine if I add the enclosing `FormGroup` that it probably should have anyway.

michioxd replied:
I found another solution, just add `CardActionArea` outside `CardContent`. Example can be found [here](https://mui.com/material-ui/react-card/#primary-action)

michioxd replied:
So, I tried to add override `CardContent` style by adding some props to `extendTheme` in `index.tsx`.

```ts
const themeExtended = experimental_extendTheme({
  colorSchemes: {
    dark: {
      palette: {
        primary: {
          main: "#000000",
        },
        secondary: {
          main: "#e65100",
        },
      },
    },
  },
  // this...
  components: {
    MuiCardContent: {
      styleOverrides: {
        root: {
          padding: '16px',
          paddingBottom: '16px !important',
        }
      }
    }
  }
});
```

Look better now:

![image](https://github.com/scottlamb/moonfire-nvr/assets/80969068/f8ff1aa1-6c22-4eee-99bf-bdd0f1eb4483)

scottlamb commented on ui/src/Live/Multiview.tsx:
```diff
@@ -106,18 +109,62 @@ function selectedReducer(old: SelectedCameras, op: SelectOp): SelectedCameras {
     searchParams.has("cams")
-      ? JSON.parse(searchParams.get("cams") || "")
+      ? JSON.parse(searchParams.get("cams") || "") :
+      localStorage.getItem("camsSelected") !== null ?
```
So this honors the URL parameters if set, and then tries local storage otherwise? Neat!

michioxd replied:
Yes!

scottlamb commented on ui/src/Live/Multiview.tsx:
```diff
@@ -106,18 +109,62 @@ function selectedReducer(old: SelectedCameras, op: SelectOp): SelectedCameras {
+    if (outerRef.current) {
+      const elem = outerRef.current;
+      //@ts-expect-error another browser (if not bruh)
+      if (document.fullscreenElement || document.webkitFullscreenElement || document.msFullscreenElement) {
```
Do we need these `webkitFullScreenElement` and `msFullScreenElement` variants? Chrome and Edge look green here: https://caniuse.com/mdn-api_document_exitfullscreen

scottlamb commented on ui/src/Live/Multiview.tsx:
```diff
@@ -165,6 +211,15 @@ const Multiview = (props: MultiviewProps) => {
     >
+      <Tooltip title="Toggle full screen">
+        <IconButton size="small" sx={{
+          position: 'fixed', background: 'rgba(255,255,255,0.3) !important', transition: '0.2s', opacity: '0.1', bottom: 10, right: 10, zIndex: 9, color: "#fff", ":hover": {
```
How would you feel about a bit more opacity? I saw this code, loaded up the interface, and still didn't see this icon the first couple times.

michioxd replied:
Maybe look better than before :)

![image](https://github.com/scottlamb/moonfire-nvr/assets/80969068/aa11f0b5-6c89-41a6-b6d9-8027e0f1a9e3)

scottlamb reviewed (commented):
Thanks! Nice improvements here, and it's a treat for me to see how other people approach UI things since I'm more of a backend developer. Some comments below.

michioxd commented on ui/src/components/ThemeMode.tsx:
```diff
@@ -0,0 +1,47 @@
+      .addEventListener("change", (e) => {
+        setDetectedSystemColorScheme(e.matches ? 'dark' : 'light');
+      });
+  }, []);
```
Nice! It's look better now.

michioxd commented on ui/src/components/ThemeMode.tsx:
```diff
@@ -0,0 +1,47 @@
+import { useColorScheme } from "@mui/material";
+import React, { createContext } from "react";
+
+interface ThemeProps {
```
I changed it to be the same as your thoughts.

michioxd commented:
What I did in this commit:
- Use your `useMediaQuery` to dynamically update the theme.
- I updated some props in `ThemeMode` and removed non-necessary props.
- Removed `webkitFullScreenElement` and `msFullScreenElement` in `Multiview` toggle full screen.
- Added a bit of opacity for the fullscreen button in `Multiview`.

scottlamb commented on ui/src/components/ThemeMode.tsx:
```diff
@@ -3,44 +3,53 @@
 }
 
 export const ThemeContext = createContext<ThemeProps>({
   currentTheme: window.matchMedia("(prefers-color-scheme: dark)").matches ? 'dark' : 'light',
```
[This comment](https://github.com/vitest-dev/vitest/issues/821#issuecomment-1046954558) looks promising for fixing this `TypeError` from `vitest`.

michioxd replied:
I see.

scottlamb commented on ui/src/components/ThemeMode.tsx:
```diff
@@ -0,0 +1,60 @@
+
+  const changeTheme = React.useCallback(() => {
+    setMode(mode === 'dark' ? 'light' : mode === 'light' ? 'system' : 'dark')
+  }, [mode]);
```
> `React Hook React.useCallback has a missing dependency: 'setMode'. Either include it or remove the dependency array`

Huh, whatever magic makes hook dependencies not need `setFoo` declared separately after `const [foo, setFoo] = useState(...)` doesn't seem to work with mui's `useColorScheme`. I guess just add `setMode` here?

scottlamb commented on ui/pnpm-lock.yaml:
```diff
@@ -0,0 +1,6048 @@
+lockfileVersion: '6.0'
```
> oh damn I used `pnpm` instead `npm` heh...

I'm not really up on the JS ecosystem. Is `pnpm` better enough to be worth switching?
* If so, we should update everything to match: instructions in `guide/build.md`, GitHub workflows, removing the `package-lock.json`
* otherwise, we should kill this file. It'll be confusing to have both, especially without anything guaranteeing they match.

scottlamb replied:
`release.yml` is still using `npm`, so it's just not locked now and the product might be something different than what was seen in any pre-release testing. We should switch it, too. I think I'd also just have `guide/install.md` say to use `pnpm` too because if you do the `npm` path, you likewise end up with something a little different than what anyone else is running with.

michioxd replied:
Hmmm, so we will mainly use `pnpm` right?

scottlamb commented:
Ci passes! 🎉 just one more comment about the npm->pnpm transition.
